src/compiler/checker.ts

```diff
@@ -37389,6 +37399,72 @@ export function createTypeChecker(host: TypeCheckerHost): TypeChecker {
         }
     }
 
+    function getTypePredicateFromBody(func: FunctionLikeDeclaration): TypePredicate | undefined {
+        switch (func.kind) {
+            case SyntaxKind.Constructor:
+            case SyntaxKind.GetAccessor:
+            case SyntaxKind.SetAccessor:
+                return undefined;
+        }
+        const functionFlags = getFunctionFlags(func);
+        if (functionFlags !== FunctionFlags.Normal || func.parameters.length === 0) return undefined;

```

Comment:
We should check the signature's `getParameterCount` instead of the raw declaration here, if possible - as is, this PR infers a predicate `f is never` for
```
const a = (...f: []) => typeof f === "undefined";
```
which, while true, is kinda funky and... probably...? not useful.

More tests with rest args are probably pertinent, too - don't see many (any?) of those in the test added. For example, while the above makes a weird predicate, this makes no predicate:
```ts
const a = (...f: ["a" | "b"]) => f[0] === "a";
```
while the non-rest version does:
```ts
const a = (f: "a" | "b") => f === "a";
```

Reply:
Good point. We should probably just ignore rest parameters since they can't be used in type predicates:

```ts
const a = (...f: ["a" | "b"]): f is ['a'] => {
    // A type predicate cannot reference a rest parameter. (1229)
    return (f[0] === "a");
}
```

[playground](https://www.staging-typescript.org/play?ts=5.5.0-pr-57465-98#code/MYewdgzgLgBAhjAvDAFAOgwMwFwwNoBEcBMAPjAQEYEC6AlLpjAJYT4Dkc7NSAfDAG8AUDFEwA9OJgBBGFACeABwCmMRQCdlAE2bA4UVXrBgQsTZmWawwVQk3Q1cdXAC2yg+rSoAjACZfAJx0ImKaUACu6mComHgADDyISRTEdADcQgC+QkA)

Is a type predicate of `x is never` ever useful? It seems more likely that it indicates some other kind of mistake.

Reply:
@weswigham I added a check for `isRestParameter` and added your two examples to the test cases. I also added a test for a function that takes a rest parameter but narrows on another, non-rest parameter.

I switched the check to `getParameterCount`. Since this requires the signature, not the declaration, I've moved it out of `getTypePredicateFromBody` and into `getTypePredicateOfSignature`.